### PR TITLE
Refactor CONTRIBUTING GUIDELINES

### PR DESCRIPTION
**Pull request by DidierRLopes**
This will provide a big improvement in our current Contributing guidelines

### CODE REVIEW

**Review by colin99d — APPROVED** [review]

**Review by minhhoang1023 — CHANGES_REQUESTED** [review]
Just reviewed the new and looks very nice overall. It has all the components needed and is split into Basic / Advanced to avoid making people feel overwhelmed.

I felt that making some minor changes to the current order would make it easier & straightforward to process. Right now, the MVC pattern was presented on top. This is a great illustration. My concern is, for contributors to fully understand it, some MVC knowledge is needed. So I would recommend moving them close to the explanation of what a View / Model/controller is.

The current View example does not follow our MVC best practices 😄 so this should be reflected.
image

Finally, the API Factory was referred several times and we don't yet have an explanation for it.

**Comment by montezdesousa** [comment]
Thanks for the review Minh!

> Just reviewed the new and looks very nice overall. It has all the components needed and is split into Basic / Advanced to avoid making people feel overwhelmed.
> 
> I felt that making some minor changes to the current order would make it easier & straightforward to process. Right now, the MVC pattern was presented on top. This is a great illustration. My concern is, for contributors to fully understand it, some MVC knowledge is needed. So I would recommend moving them close to the explanation of what a View / Model/controller is.

I agree with you on the MVC, so you are suggesting we start with "Follow Coding Guidelines" straight away? My issue with this would be similar. Coding guidelines require knowledge about code structure that is not there before. As an alternative we could pull "Adding a new command" to be the first topic. Idk, what do you think should be the order?


> The current View example does not follow our MVC best practices 😄 so this should be reflected. image
🤦 

> Finally, the API Factory was referred several times and we don't yet have an explanation for it.

Should we create a dedicated topic for it?

**Comment by minhhoang1023** [comment]
> I agree with you on the MVC, so you are suggesting we start with "Follow Coding Guidelines" straight away? My issue with this would be similar. Coding guidelines require knowledge about code structure that is not there before. As an alternative we could pull "Adding a new command" to be the first topic. Idk, what do you think should be the order?
> 
> >

I have an idea - let's chat together on Monday!

**Comment by montezdesousa** [comment]
> Thanks for the review Minh!
> 
> > Just reviewed the new and looks very nice overall. It has all the components needed and is split into Basic / Advanced to avoid making people feel overwhelmed.
> > I felt that making some minor changes to the current order would make it easier & straightforward to process. Right now, the MVC pattern was presented on top. This is a great illustration. My concern is, for contributors to fully understand it, some MVC knowledge is needed. So I would recommend moving them close to the explanation of what a View / Model/controller is.
> 
> I agree with you on the MVC, so you are suggesting we start with "Follow Coding Guidelines" straight away? My issue with this would be similar. Coding guidelines require knowledge about code structure that is not there before. As an alternative we could pull "Adding a new command" to be the first topic. Idk, what do you think should be the order?
> 
> > The current View example does not follow our MVC best practices 😄 so this should be reflected. image
> > 🤦
> 
> > Finally, the API Factory was referred several times and we don't yet have an explanation for it.
> 
> Should we create a dedicated topic for it?

Ok so I moved "Adding a new command" up, updated the `get_most_shorted` function and created a new work in progress topic for the API structure.

File: CONTRIBUTING.md
Comment (DidierRLopes) [inline CONTRIBUTING.md:883]: Mention here that we want to accept date time in string so that is easier to deal from user perspective. And then we convert to date time and check if the date is valid within the function

File: CONTRIBUTING.md
Comment (DidierRLopes) [inline CONTRIBUTING.md:1381]: @Chavithra can you take this one? What should we add for the contributor to know where to add this?

File: CONTRIBUTING.md
Comment (DidierRLopes) [inline CONTRIBUTING.md:1411]: Change to and add straight after INTERNATIONALIZATION title

```
WORK IN PROGRESS - The menu can be internationalised BUT we do not support yet help commands`-h` internationalization.
```

File: CONTRIBUTING.md
Comment (DidierRLopes) [inline CONTRIBUTING.md:1451]: @Chavithra I still find tests one of the hardest things to do on the terminal - and have gotten that feedback multiple times. Can we write more and better documentation around it?

File: CONTRIBUTING.md
Comment (DidierRLopes) [inline CONTRIBUTING.md:1472]: Probably remove this if we will have an API readme, and this is about usability?

**Review by DidierRLopes — COMMENTED** [review]
Api -> Wrapper API (also api needs to be capitalised)
Back-end -> backend and front-end -> frontend

I added my main points to this document, but it's looking SO MUCH BETTER. Nice work.

File: CONTRIBUTING.md
Comment (minhhoang1023) [reply CONTRIBUTING.md:1451]: @DidierRLopes Got the same feelings. Chavi and I talked about this and I added some more explanations on tests a while ago to the README for tests. We could link it here as it's such an extensive topic

Comment (DidierRLopes) [reply CONTRIBUTING.md:1451]: Am fine with that. But definitely think we need to drill more on that topic. A lot of devs are a bit scared by tests.

File: CONTRIBUTING.md
Comment (montezdesousa) [reply CONTRIBUTING.md:1469]: We remove the topic? Right now yes, that document will be improved later

File: CONTRIBUTING.md
Comment (montezdesousa) [reply CONTRIBUTING.md:880]: Done

File: CONTRIBUTING.md
Comment (montezdesousa) [reply CONTRIBUTING.md]: Is it better now?

File: CONTRIBUTING.md
Comment (montezdesousa) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md
Comment (montezdesousa) [reply CONTRIBUTING.md]: Is it better rn? Not sure if I got it right

File: CONTRIBUTING.md
Comment (montezdesousa) [reply CONTRIBUTING.md:98]: Example for adding a new command right?

**Comment by DidierRLopes** [comment]
Thanks @montezdesousa! This is starting to look very nice! I can't approve this PR because I created it, but @minhhoang1023 feel free to approve it when you think is good :)

**Review by piiq — APPROVED** [review]

**Comment by piiq** [comment]
@minhhoang1023's change request is blocking me from merging. Is there anything else required here?

**Comment by minhhoang1023** [comment]
@piiq The logging section is currently empty. Once @Chavithra adds it, lgtm for merging

File: CONTRIBUTING.md
Comment (minhhoang1023) [reply CONTRIBUTING.md:1406]: @Chavithra

**Review by minhhoang1023 — APPROVED** [review]

**Pull request merged**
